<font face="Times" size=13 color=#A52A2A >
Fetch - Data Analyst Take Home

<font face="Times" size=4 color=#121211 >
    Name: JieXin Chen
    <br>January 22, 2025  

In my analysis process, I import CSV files into Excel for closer inspection. Excel's filter function is helpful for quickly identifying empty values or anomalies in the data.

In [307]:
# Import necessary libraries
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
import seaborn as sns
from datetime import datetime
import numpy as np

# Step 1: Load the datasets 

In this step, I will load the provided CSV files into Python using pandas. This allows me to structure the data into DataFrames for analysis. 

In [308]:
# Load Users, Products, and Transactions dataset
users_path = "/Users/jc/Desktop/xing/fetch/interview/USER_TAKEHOME.csv"  
transactions_path = "/Users/jc/Desktop/xing/fetch/interview/TRANSACTION_TAKEHOME.csv" 
products_path = "/Users/jc/Desktop/xing/fetch/interview/PRODUCTS_TAKEHOME.csv" 

# Load data into pandas DataFrames
users_df = pd.read_csv(users_path, dtype=str)
transactions_df = pd.read_csv(transactions_path, dtype=str)
products_df = pd.read_csv(products_path, dtype=str, encoding="utf-8")

# Step 2: Overview of each dataset

I will inspect the first few rows of each dataset to get an overview of the data structure. This helps me understand the data type in each column and identify any immediate issues, such as unexpected formats or missing data.



In [309]:
print("Users Dataset:")
users_df.head()

Users Dataset:


ID               CREATED_DATE  \
0  5ef3b4f17053ab141787697d  2020-06-24 20:17:54.000 Z   
1  5ff220d383fcfc12622b96bc  2021-01-03 19:53:55.000 Z   
2  6477950aa55bb77a0e27ee10  2023-05-31 18:42:18.000 Z   
3  658a306e99b40f103b63ccf8  2023-12-26 01:46:22.000 Z   
4  653cf5d6a225ea102b7ecdc2  2023-10-28 11:51:50.000 Z   

                  BIRTH_DATE STATE LANGUAGE  GENDER  
0  2000-08-11 00:00:00.000 Z    CA   es-419  female  
1  2001-09-24 04:00:00.000 Z    PA       en  female  
2  1994-10-28 00:00:00.000 Z    FL   es-419  female  
3                        NaN    NC       en     NaN  
4  1972-03-19 00:00:00.000 Z    PA       en  female

In [310]:
print("Transactions Dataset:")
transactions_df.head()

Transactions Dataset:


RECEIPT_ID PURCHASE_DATE  \
0  0000d256-4041-4a3e-adc4-5623fb6e0c99    2024-08-21   
1  0001455d-7a92-4a7b-a1d2-c747af1c8fd3    2024-07-20   
2  00017e0a-7851-42fb-bfab-0baa96e23586    2024-08-18   
3  000239aa-3478-453d-801e-66a82e39c8af    2024-06-18   
4  00026b4c-dfe8-49dd-b026-4c2f0fd5c6a1    2024-07-04   

                   SCAN_DATE STORE_NAME                   USER_ID  \
0  2024-08-21 14:19:06.539 Z    WALMART  63b73a7f3d310dceeabd4758   
1  2024-07-20 09:50:24.206 Z       ALDI  62c08877baa38d1a1f6c211a   
2  2024-08-19 15:38:56.813 Z    WALMART  60842f207ac8b7729e472020   
3  2024-06-19 11:03:37.468 Z  FOOD LION  63fcd7cea4f8442c3386b589   
4  2024-07-05 15:56:43.549 Z   RANDALLS  6193231ae9b3d75037b0f928   

        BARCODE FINAL_QUANTITY FINAL_SALE  
0  015300014978           1.00             
1           NaN           zero       1.49  
2  078742229751           1.00             
3  783399746536           zero       3.49  
4  047900501183           1.00

Based on the dataset, I identified an issue in the FINAL_SALE column where some values were blank or invalid. Since this column is expected to contain numerical data representing sales values, such inconsistencies needed to be addressed. To resolve this, I used the pd.to_numeric() function to convert the FINAL_SALE column into a numeric data type. This method automatically converted invalid entries into NaN, allowing them to be appropriately treated as missing values during further analysis or imputation. This ensures the column is now clean and ready for numerical operations.

In [311]:
# Convert the FINAL_SALE column to numeric, coercing invalid entries to NaN
transactions_df['FINAL_SALE'] = pd.to_numeric(transactions_df['FINAL_SALE'], errors='coerce')

# Check the data type of the FINAL_SALE column
print(transactions_df['FINAL_SALE'].dtypes)

float64


In [312]:
# Display the first few rows of the transactions dataset again to verify changes
# Ensures that the 'FINAL_SALE' column has been successfully converted to numeric
transactions_df.head()

RECEIPT_ID PURCHASE_DATE  \
0  0000d256-4041-4a3e-adc4-5623fb6e0c99    2024-08-21   
1  0001455d-7a92-4a7b-a1d2-c747af1c8fd3    2024-07-20   
2  00017e0a-7851-42fb-bfab-0baa96e23586    2024-08-18   
3  000239aa-3478-453d-801e-66a82e39c8af    2024-06-18   
4  00026b4c-dfe8-49dd-b026-4c2f0fd5c6a1    2024-07-04   

                   SCAN_DATE STORE_NAME                   USER_ID  \
0  2024-08-21 14:19:06.539 Z    WALMART  63b73a7f3d310dceeabd4758   
1  2024-07-20 09:50:24.206 Z       ALDI  62c08877baa38d1a1f6c211a   
2  2024-08-19 15:38:56.813 Z    WALMART  60842f207ac8b7729e472020   
3  2024-06-19 11:03:37.468 Z  FOOD LION  63fcd7cea4f8442c3386b589   
4  2024-07-05 15:56:43.549 Z   RANDALLS  6193231ae9b3d75037b0f928   

        BARCODE FINAL_QUANTITY  FINAL_SALE  
0  015300014978           1.00         NaN  
1           NaN           zero        1.49  
2  078742229751           1.00         NaN  
3  783399746536           zero        3.49  
4  047900501183           1.00         NaN

In [313]:
print("Products Dataset:")
products_df.head()

Products Dataset:


CATEGORY_1              CATEGORY_2                   CATEGORY_3  \
0  Health & Wellness           Sexual Health  Conductivity Gels & Lotions   
1             Snacks           Puffed Snacks         Cheese Curls & Puffs   
2  Health & Wellness               Hair Care        Hair Care Accessories   
3  Health & Wellness               Oral Care                   Toothpaste   
4  Health & Wellness  Medicines & Treatments               Essential Oils   

  CATEGORY_4                                       MANUFACTURER  \
0        NaN                                                NaN   
1        NaN                                                NaN   
2        NaN                           PLACEHOLDER MANUFACTURER   
3        NaN                                  COLGATE-PALMOLIVE   
4        NaN  MAPLE HOLISTICS AND HONEYDEW PRODUCTS INTERCHA...   

             BRAND        BARCODE  
0              NaN   796494407820  
1              NaN   023278011028  
2          ELECSOP   461817824225  
3          COLGATE   035000466815  
4  MAPLE HOLISTICS  0806810850459

# Step 3: Inspect and Define Data Types
This step will both retrieve the column names and data types, and also define/standardize the correct data types based on observations.

In [314]:
print("\nUsers Dataset Info:")
users_df.info()


Users Dataset Info:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 6 columns):
 #   Column        Non-Null Count   Dtype 
---  ------        --------------   ----- 
 0   ID            100000 non-null  object
 1   CREATED_DATE  100000 non-null  object
 2   BIRTH_DATE    96325 non-null   object
 3   STATE         95188 non-null   object
 4   LANGUAGE      69492 non-null   object
 5   GENDER        94108 non-null   object
dtypes: object(6)
memory usage: 4.6+ MB


In [315]:
print("\nTransactions Dataset Info:")
transactions_df.info()


Transactions Dataset Info:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 8 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   RECEIPT_ID      50000 non-null  object 
 1   PURCHASE_DATE   50000 non-null  object 
 2   SCAN_DATE       50000 non-null  object 
 3   STORE_NAME      50000 non-null  object 
 4   USER_ID         50000 non-null  object 
 5   BARCODE         44238 non-null  object 
 6   FINAL_QUANTITY  50000 non-null  object 
 7   FINAL_SALE      37500 non-null  float64
dtypes: float64(1), object(7)
memory usage: 3.1+ MB


In [316]:
print("\nProducts Dataset Info:")
products_df.info()


Products Dataset Info:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 845552 entries, 0 to 845551
Data columns (total 7 columns):
 #   Column        Non-Null Count   Dtype 
---  ------        --------------   ----- 
 0   CATEGORY_1    845441 non-null  object
 1   CATEGORY_2    844128 non-null  object
 2   CATEGORY_3    784986 non-null  object
 3   CATEGORY_4    67459 non-null   object
 4   MANUFACTURER  619078 non-null  object
 5   BRAND         619080 non-null  object
 6   BARCODE       841527 non-null  object
dtypes: object(7)
memory usage: 45.2+ MB


In [317]:
# Code for Defining Data Types
# Define data types for Users dataset
users_df['ID'] = users_df['ID'].astype('string')
users_df['CREATED_DATE'] = pd.to_datetime(users_df['CREATED_DATE'], errors='coerce')
users_df['BIRTH_DATE'] = pd.to_datetime(users_df['BIRTH_DATE'], errors='coerce')
users_df['STATE'] = users_df['STATE'].astype('string')
users_df['LANGUAGE'] = users_df['LANGUAGE'].astype('string')
users_df['GENDER'] = users_df['GENDER'].astype('string')

# Define data types for Transactions dataset
transactions_df['RECEIPT_ID'] = transactions_df['RECEIPT_ID'].astype('string')
transactions_df['PURCHASE_DATE'] = pd.to_datetime(transactions_df['PURCHASE_DATE'], errors='coerce')
transactions_df['SCAN_DATE'] = pd.to_datetime(transactions_df['SCAN_DATE'], errors='coerce')
transactions_df['STORE_NAME'] = transactions_df['STORE_NAME'].astype('string')
transactions_df['USER_ID'] = transactions_df['USER_ID'].astype('string')
transactions_df['BARCODE'] = transactions_df['BARCODE'].astype('string')
transactions_df['FINAL_SALE'] = transactions_df['FINAL_SALE'].astype('float64')
# Replace 'zero' with 0 in FINAL_QUANTITY and then convert to float64
transactions_df['FINAL_QUANTITY'] = transactions_df['FINAL_QUANTITY'].replace('zero', 0)
transactions_df['FINAL_QUANTITY'] = pd.to_numeric(transactions_df['FINAL_QUANTITY'], errors='coerce')

# Define data types for Products dataset
products_df['CATEGORY_1'] = products_df['CATEGORY_1'].astype('string')
products_df['CATEGORY_2'] = products_df['CATEGORY_2'].astype('string')
products_df['CATEGORY_3'] = products_df['CATEGORY_3'].astype('string')
products_df['CATEGORY_4'] = products_df['CATEGORY_4'].astype('string')
products_df['MANUFACTURER'] = products_df['MANUFACTURER'].astype('string')
products_df['BRAND'] = products_df['BRAND'].astype('string')
products_df['BARCODE'] = products_df['BARCODE'].astype('string')

# Step 4: Check for missing values in each dataset

I will calculate the percentage of missing values for each column in the datasets. This helps identify columns with a high proportion of missing data, which may need to be addressed during data cleaning.

In [318]:
print("Users Missing Values:")
users_df.isnull().sum() / len(users_df) * 100

Users Missing Values:


ID               0.000
CREATED_DATE     0.000
BIRTH_DATE       3.675
STATE            4.812
LANGUAGE        30.508
GENDER           5.892
dtype: float64



__Summary of Missing Data in the Users Dataset__
1. BIRTH_DATE:
- 3.68% of the entries are missing.
- This may affect age-related analyses if the missing values are not addressed.
2. STATE:
- 4.81% of the entries are missing.
- This could impact regional analysis or segmentation of users by location.
3. LANGUAGE:
- A significant 30.51% of the entries are missing.
- This large percentage may make it difficult to analyze user preferences based on language.
4. GENDER:
- 5.89% of the entries are missing.
- Missing gender data could affect demographic segmentation.
5. ID and CREATED_DATE:
- Both columns have complete data with no missing values, which ensures that unique user identification and account creation dates are reliable.

In [319]:
print("Transactions Missing Values:")
transactions_df.isnull().sum() / len(transactions_df) * 100

Transactions Missing Values:


RECEIPT_ID         0.000
PURCHASE_DATE      0.000
SCAN_DATE          0.000
STORE_NAME         0.000
USER_ID            0.000
BARCODE           11.524
FINAL_QUANTITY     0.000
FINAL_SALE        25.000
dtype: float64

__Summary of Missing Data in the Transactions Dataset__
1. BARCODE:
- 11.52% of the entries are missing.
- This poses a notable challenge since the BARCODE column is essential for linking transactions to specific products.
- The absence of values in this column may limit detailed product-level analysis and hinder integration with the products dataset.
2. Other Columns:
- The columns RECEIPT_ID, PURCHASE_DATE, SCAN_DATE, STORE_NAME, and USER_ID have complete data with 0% missing values, ensuring consistency for transaction-level analyses, such as examining sales trends, user behaviors, and store performance.
3. FINAL_QUANTITY and FINAL_SALE:
- While FINAL_QUANTITY has no missing values, the FINAL_SALE column has 25% missing data after correcting its format to numeric.
- These missing entries could impact revenue-related analysis and need to be addressed, either by imputation or exclusion, depending on business priorities.

In [320]:
print("Products Missing Values:")
products_df.isnull().sum() / len(products_df) * 100

Products Missing Values:


CATEGORY_1       0.013128
CATEGORY_2       0.168411
CATEGORY_3       7.162895
CATEGORY_4      92.021898
MANUFACTURER    26.784160
BRAND           26.783923
BARCODE          0.476020
dtype: float64

__Summary of Missing Data in the Products Dataset__
1. CATEGORY_1 to CATEGORY_4:
- Missing data percentages:
CATEGORY_1: 0.01%
CATEGORY_2: 0.17%
CATEGORY_3: 7.16%
CATEGORY_4: 92.02%
- These columns represent a hierarchical structure of product categorization, starting from broader categories (CATEGORY_1) to increasingly detailed subcategories (CATEGORY_4).
- The high missingness in CATEGORY_4 is expected, as not all products require deep levels of categorization. 
- This pattern reflects the natural structure of product data rather than a quality issue.
2. MANUFACTURER and BRAND:
- Both columns have approximately 26.78% missing values.
- Missing data here may impact analyses involving brand-specific or manufacturer-specific insights. 
- It is essential to consider strategies such as imputing missing values for critical use cases or focusing analyses on non-missing data.
3. BARCODE:
- Missing data percentage: 0.48%.
- BARCODE is critical for linking products with transactions. 
- Although the missing percentage is low, efforts should be made to ensure that missing values do not disrupt the integration of datasets or analyses at the product level.


# Step 5: Descriptive Statistics for Categorical and Other Columns
I will generate descriptive statistics for all columns in the dataset, including counts, unique values, most frequent values, and their frequencies. This provides insights into the distribution of categorical data and helps identify any inconsistencies or anomalies that require further cleaning or analysis.

In [321]:
print("\nUsers Dataset Summary Statistics:")
users_df.describe(include="all")


Users Dataset Summary Statistics:


ID                         CREATED_DATE  \
count                     100000                               100000   
unique                    100000                                  NaN   
top     5ef3b4f17053ab141787697d                                  NaN   
freq                           1                                  NaN   
mean                         NaN  2022-01-07 05:31:20.864859648+00:00   
min                          NaN            2014-04-18 23:14:55+00:00   
25%                          NaN            2020-12-01 21:16:19+00:00   
50%                          NaN            2022-03-07 01:03:37+00:00   
75%                          NaN     2023-01-30 13:47:44.500000+00:00   
max                          NaN            2024-09-11 17:59:15+00:00   

                                 BIRTH_DATE  STATE LANGUAGE  GENDER  
count                                 96325  95188    69492   94108  
unique                                  NaN     52        2      11  
top                                     NaN     TX       en  female  
freq                                    NaN   9028    63403   64240  
mean    1984-09-02 02:39:04.710417920+00:00    NaN      NaN     NaN  
min               1900-01-01 00:00:00+00:00    NaN      NaN     NaN  
25%               1974-03-04 00:00:00+00:00    NaN      NaN     NaN  
50%               1985-10-25 00:00:00+00:00    NaN      NaN     NaN  
75%               1998-02-02 05:00:00+00:00    NaN      NaN     NaN  
max               2022-04-03 07:00:00+00:00    NaN      NaN     NaN

__The potential issues identified in the descriptive statistics of the users dataset:__
1. Unrealistic Data
- BIRTH_DATE:
  - The minimum value 1900-01-01 is highly unrealistic, implying users aged 123 years or older.
  - This is likely due to placeholder values or data entry errors and should be flagged for correction or exclusion.
- CREATED_DATE:
  - The range of values (2014 to 2024) appears reasonable and aligns with expected account creation timelines.
2. Inconsistent Categories
- GENDER:
  - The column contains 11 unique categories, including not_listed, other, and prefer_not_to_say.
  - These categories may overlap or lack clear differentiation, making analysis more complex.
  - Logical grouping of similar categories (e.g., combining into broader groups) is recommended to streamline analysis.
3. Frequency Issues
- STATE:
  - The state TX has a disproportionately high frequency, with 9,028 entries.
  - This could indicate an uneven geographic distribution or sampling bias.
  - Further investigation is necessary to understand if this represents the actual data distribution or a collection anomaly.

In [322]:
print("\nTransactions Dataset Summary Statistics:")
transactions_df.describe(include="all")


Transactions Dataset Summary Statistics:


RECEIPT_ID               PURCHASE_DATE  \
count                                  50000                       50000   
unique                                 24440                         NaN   
top     bedac253-2256-461b-96af-267748e6cecf                         NaN   
freq                                      12                         NaN   
mean                                     NaN  2024-07-24 09:44:17.664000   
min                                      NaN         2024-06-12 00:00:00   
25%                                      NaN         2024-07-03 00:00:00   
50%                                      NaN         2024-07-24 00:00:00   
75%                                      NaN         2024-08-15 00:00:00   
max                                      NaN         2024-09-08 00:00:00   
std                                      NaN                         NaN   

                                  SCAN_DATE STORE_NAME  \
count                                 50000      50000   
unique                                  NaN        954   
top                                     NaN    WALMART   
freq                                    NaN      21326   
mean    2024-07-27 00:29:55.204816896+00:00        NaN   
min        2024-06-12 06:36:34.910000+00:00        NaN   
25%     2024-07-05 10:59:32.364750080+00:00        NaN   
50%     2024-07-26 11:43:59.964499968+00:00        NaN   
75%     2024-08-17 17:10:10.855000064+00:00        NaN   
max        2024-09-08 23:07:19.836000+00:00        NaN   
std                                     NaN        NaN   

                         USER_ID       BARCODE  FINAL_QUANTITY    FINAL_SALE  
count                      50000         44238    50000.000000  37500.000000  
unique                     17694         11028             NaN           NaN  
top     64e62de5ca929250373e6cf5  078742223759             NaN           NaN  
freq                          22           182             NaN           NaN  
mean                         NaN           NaN        0.814427      4.576384  
min                          NaN           NaN        0.000000      0.000000  
25%                          NaN           NaN        0.007500      1.790000  
50%                          NaN           NaN        1.000000      3.000000  
75%                          NaN           NaN        1.000000      5.190000  
max                          NaN           NaN      276.000000    462.820000  
std                          NaN           NaN        1.841794      6.625663

__The potential issues identified in the descriptive statistics of the transaction dataset:__
1. Missing Values (NaN):
- BARCODE: Approximately 11.5% of the BARCODE column contains missing values. Since this column is critical for product identification, these missing entries may affect analyses or integrations that rely on product-level data.
- FINAL_SALE: Around 25% of the FINAL_SALE column contains missing values. This could significantly impact revenue-related analyses, as these rows lack information about transaction totals.
2. Issues with USER_ID and RECEIPT_ID:
- USER_ID:
  - The dataset contains 50,000 rows but only 17,694 unique USER_ID values. This suggests that some users have multiple transactions, which is expected for repeat customers. However, this may lead to potential bias in user-level analyses unless handled appropriately.
- RECEIPT_ID:
  - Only 24,440 unique RECEIPT_ID values exist in the dataset, despite 50,000 rows. This indicates that many receipt IDs are duplicated, which might suggest:
  - Errors in the data entry or duplication of transactions.
  - Bulk purchases recorded under the same receipt ID.
  - Transactions split into multiple rows for the same receipt.
This requires further investigation to understand whether the duplication is intentional or a data quality issue.

In [323]:
print("\nProducts Dataset Summary Statistics:")
products_df.describe(include="all")


Products Dataset Summary Statistics:


CATEGORY_1 CATEGORY_2        CATEGORY_3 CATEGORY_4  \
count              845441     844128            784986      67459   
unique                 27        121               344        127   
top     Health & Wellness      Candy  Confection Candy  Lip Balms   
freq               512695     121036             56965       9737   

                    MANUFACTURER      BRAND       BARCODE  
count                     619078     619080        841527  
unique                      4354       8122        841525  
top     PLACEHOLDER MANUFACTURER  REM BRAND  017000329260  
freq                       86902      20813             2

__The potential issues in the products dataset:__
1. Incomplete Data Across Categories

- CATEGORY_2, CATEGORY_3, CATEGORY_4:
The number of entries decreases significantly from CATEGORY_2 (844,128) to CATEGORY_4 (67,459).
- This reflects a hierarchical structure where deeper categorizations may not apply to all products. While this is expected, it should be validated to ensure consistent and accurate data representation.
2. High Frequency in Specific Categories
- CATEGORY_1:
The "Health & Wellness" category contains 512,695 entries, dominating the dataset. This unbalanced distribution could introduce bias into the analysis or insights derived from the data.
3. High Variability in Categorical Data
- CATEGORY_3 and CATEGORY_4:
CATEGORY_3 has 344 unique values, and CATEGORY_4 has 127 unique values. This high variability may complicate grouping, aggregation, or analysis, requiring careful categorization or feature engineering.
4. Irregularities in Barcodes

- Duplicate Barcodes:
Among 841,527 total entries, only 841,342 barcodes are unique, indicating some duplicates.
Potential problems with these duplicates include:
  - Data entry errors or placeholder values.
  - Multiple records representing the same product in different contexts.

# Step 6: Check for duplicates in each dataset
I will check each dataset for duplicate rows, as these can affect the accuracy of analysis and may need to be removed.

In [324]:
print("\nDuplicate Rows Analysis:")
print(f"Users duplicate rows: {users_df.duplicated().sum()}")
print(f"Transactions duplicate rows: {transactions_df.duplicated().sum()}")
print(f"Products duplicate rows: {products_df.duplicated().sum()}")


Duplicate Rows Analysis:
Users duplicate rows: 0
Transactions duplicate rows: 171
Products duplicate rows: 57


In [325]:
duplicates_transactions = transactions_df[transactions_df.duplicated()]
print(duplicates_transactions.head())
duplicates_products = products_df[products_df.duplicated()]
print(duplicates_products.head())

                                RECEIPT_ID PURCHASE_DATE  \
2724  1c12bd8a-c68c-41ee-a26d-294021d3e0b8    2024-09-07   
4192  2acd7e8d-37df-4e51-8ee5-9a9c8c1d9711    2024-09-08   
6572  431fe612-ed55-470e-939c-043ad31f33f3    2024-09-07   
6623  43955b35-6fbc-4909-a4de-1a0de0dc387f    2024-09-06   
7675  4ec870d2-c39f-4a40-bf8a-26a079409b20    2024-09-08   

                            SCAN_DATE            STORE_NAME  \
2724 2024-09-07 15:44:35.241000+00:00                   CVS   
4192 2024-09-08 11:13:01.935000+00:00               WALMART   
6572 2024-09-07 16:39:01.409000+00:00  DOLLAR GENERAL STORE   
6623 2024-09-08 18:32:30.031000+00:00               WALMART   
7675 2024-09-08 19:39:01.589000+00:00               WALMART   

                       USER_ID       BARCODE  FINAL_QUANTITY  FINAL_SALE  
2724  65af09757050d0a6206ab136          <NA>             1.0         NaN  
4192  663140f9b7b24d45d938f3be  024000048336             1.0         NaN  
6572  5e038cebcb322c11de193bb7  012

# Step 7: Clean Data for EDA Preparation
In this step, I address and resolve the data quality issues identified earlier, including handling duplicate rows, correcting inconsistent values, and ensuring logical integrity in the dataset. These cleaning actions aim to prepare the data for effective exploratory data analysis (EDA) by ensuring its accuracy and consistency.

__Filter out rows where the 'BARCODE' column has missing values__

I deleted the product and transaction records with null values in the BARCODE field because it serves as a link to the transaction data. Without a BARCODE, those records are not useful.

In [326]:
products_df = products_df[products_df['BARCODE'].notnull()]
print(f"Remaining rows in products_df: {len(products_df)}")
# Check the duplicate dataset
print(f"Products duplicate rows: {products_df.duplicated().sum()}")

Remaining rows in products_df: 841527
Products duplicate rows: 0


In [327]:
transactions_df = transactions_df[transactions_df['BARCODE'].notnull()]
print(f"Remaining rows in transactions_df: {len(transactions_df)}")

Remaining rows in transactions_df: 44238


This step ensures only products with valid barcodes remain, as the 'BARCODE' is used to connect to the transactions table.

__Removing Incomplete and Duplicate Transactions__

This section filters out rows where either FINAL_QUANTITY or FINAL_SALE is missing or zero, ensuring the dataset retains only valid transaction records. Duplicate RECEIPT_ID entries are also removed, keeping only the first occurrence to maintain data integrity.

In [328]:
# Filter the dataset to remove rows where FINAL_QUANTITY or FINAL_SALE is 0 or blank
transactions_df = transactions_df[
    (transactions_df['FINAL_QUANTITY'] != 0) & 
    (transactions_df['FINAL_SALE'] != 0)
]

# Drop duplicate RECEIPT_ID while keeping the first occurrence (or last depending on preference)
transactions_df = transactions_df.sort_values(by=['FINAL_SALE'], ascending=False)
transactions_df = transactions_df.drop_duplicates(subset=['RECEIPT_ID'], keep='first')

# Check the resulting dataset
print(f"Transactions duplicate rows: {transactions_df.duplicated().sum()}")

Transactions duplicate rows: 0


In the transactions dataset, I identified that some BARCODE values were "-1", which is not logical for barcodes. Therefore, I removed these invalid values by setting them to empty strings to ensure the integrity of the dataset.

In [329]:
# Ensure the BARCODE column is treated as a string
transactions_df['BARCODE'] = transactions_df['BARCODE'].astype(str)

# Replace '-1' with an empty string
transactions_df['BARCODE'] = transactions_df['BARCODE'].apply(lambda x: '' if x == '-1' else x)


__Cleaning and Standardizing the GENDER Column__

This section involves defining a mapping function to clean and standardize the values in the GENDER column, addressing inconsistencies, handling missing values, and grouping less common responses into a logical framework.

In [330]:
# Display all unique values in the GENDER column
users_df['GENDER'].unique()

<StringArray>
[                'female',                     <NA>,                   'male',
             'non_binary',            'transgender',      'prefer_not_to_say',
             'not_listed',             'Non-Binary',                'unknown',
          'not_specified', "My gender isn't listed",      'Prefer not to say']
Length: 12, dtype: string

In [331]:
# Define a mapping function for cleaning the GENDER column
def clean_gender(gender):
    if pd.isnull(gender):
        return 'other'  # Treat NaN as 'other'
    gender = gender.lower().strip()
    if gender in ['female']:
        return 'female'
    elif gender in ['male']:
        return 'male'
    elif gender in ['non_binary', 'non-binary']:
        return 'non_binary'
    elif gender in ['prefer_not_to_say', 'prefer not to say']:
        return 'prefer_not_to_say'
    elif gender in ['transgender']:
        return 'transgender'
    elif gender in ['not_listed', "my gender isn't listed"]:
        return 'not_listed'
    elif gender in ['not_specified', 'unknown']:
        return 'other'
    else:
        return 'other'  # Default to 'other' for any unexpected values

# Apply the mapping function and replace the GENDER column
users_df['GENDER'] = users_df['GENDER'].apply(clean_gender)

# Display the unique values in the updated GENDER column
print(users_df['GENDER'].value_counts())

GENDER
female               64240
male                 25829
other                 6116
transgender           1772
prefer_not_to_say     1351
non_binary             507
not_listed             185
Name: count, dtype: int64


__The GENDER column was cleaned and consolidated to simplify analysis while maintaining logical distinctions.__

Here’s the explanation:
- Purpose: Reduce the number of unique values for easier interpretation and processing.
- Why Keep not_listed:
  It reflects users indicating their identity wasn’t included in the options, which differs from ambiguous entries like unknown.
- Why Group into other:
  - Ambiguity: Terms like not_specified and unknown don’t indicate any clear gender identity but rather imply missing or incomplete information. Grouping these into other avoids overcomplicating the dataset while maintaining logical consistency.
  - Practicality: These terms don’t provide actionable insights individually, so grouping them ensures clarity without diminishing analytical value.
- Final Categories:
  - female and male: Standard binary categories.
  - non_binary and transgender: Preserved for inclusivity.
  - not_listed: Retained to reflect users who felt their gender wasn’t represented.
  - prefer_not_to_say: Indicates intentional non-disclosure.
  - other: Includes ambiguous or undefined entries like unknown, not_specified, and NaN.

__Filter Out Rows with Unreasonable Ages__

I defined a reasonable age as below 100 for this project. In a formal project, this criteria would be confirmed with the responsible personnel to ensure alignment.

In [332]:
# Calculate age
today = datetime.today()
users_df['AGE'] = users_df['BIRTH_DATE'].apply(
    lambda x: today.year - x.year - ((today.month, today.day) < (x.month, x.day)) if pd.notnull(x) else None
)
# Check for unreasonable ages
users_df['UNREASONABLE_AGE'] = users_df['AGE'].apply(
    lambda x: x < 0 or x > 100 if x is not None else None
)

# Display rows with unreasonable ages
UNREASONABLE_AGE = users_df[users_df['UNREASONABLE_AGE'] == True]
UNREASONABLE_AGE

ID              CREATED_DATE  \
3656   62be5974baa38d1a1f6b6725 2022-07-01 02:18:28+00:00   
5190   5f20a5a2b06a3214c6db4811 2020-07-28 22:24:34+00:00   
5878   60ac6acc79ed9200a6ebc24a 2021-05-25 03:11:08+00:00   
6499   60a6e4af3369535cb6c4c89d 2021-05-20 22:37:35+00:00   
7723   624501294eadbb0ccddf2cc6 2022-03-31 01:17:29+00:00   
...                         ...                       ...   
96783  5ee14ae9e7bf80140949615c 2020-06-10 21:04:43+00:00   
98338  6418e8d1197cb65cc272db79 2023-03-20 23:14:25+00:00   
98639  61a1127207acef7276e6158f 2021-11-26 16:59:30+00:00   
98847  60bc0f11a3100c10461d86fa 2021-06-05 23:56:01+00:00   
98873  5ca8bf5c5f04d512fc5d3dc6 2019-04-06 15:01:48+00:00   

                     BIRTH_DATE STATE LANGUAGE       GENDER    AGE  \
3656  1903-01-01 05:00:00+00:00    PA       en         male  122.0   
5190  1923-07-28 04:00:00+00:00    PA     <NA>  transgender  101.0   
5878  1901-05-25 00:00:00+00:00    UT     <NA>       female  123.0   
6499  1901-10-13 00:00:00+00:00    CA     <NA>         male  123.0   
7723  1922-12-22 00:00:00+00:00    IL   es-419         male  102.0   
...                         ...   ...      ...          ...    ...   
96783 1907-11-13 06:00:00+00:00    LA       en         male  117.0   
98338 1906-01-01 06:00:00+00:00    TN       en         male  119.0   
98639 1901-01-01 05:00:00+00:00    PA       en         male  124.0   
98847 1901-01-01 05:00:00+00:00    NY     <NA>       female  124.0   
98873 1917-12-07 00:00:00+00:00    PA     <NA>       female  107.0   

       UNREASONABLE_AGE  
3656               True  
5190               True  
5878               True  
6499               True  
7723               True  
...                 ...  
96783              True  
98338              True  
98639              True  
98847              True  
98873              True  

[61 rows x 8 columns]

__Replace Unreasonable Ages with NaN__

To maintain data integrity, I replaced rows with ages above the defined limit (100 years) with NaN for further analysis or review.

In [333]:
# Replace unreasonable ages with NaN
users_df.loc[users_df['UNREASONABLE_AGE'], 'BIRTH_DATE'] = None

- I set the reasonable age range to 0-100 to account for living individuals, acknowledging that accounts for young children might be created by their guardians.
- Ages over 100 are unlikely and often result from data entry errors.
- I replaced unreasonable values with blanks rather than removing rows to preserve other valid data.
- I recommend adding input validation during account registration, such as limiting birth year entry to a realistic range (e.g., 1920 to the current year).

__Standardize Date Format for Tableau Calculations__

The CREATED_DATE and BIRTH_DATE columns were reformatted to the YYYY-MM-DD standard, making them suitable for calculations and compatibility with Tableau.



In [334]:
# Convert CREATED_DATE and BIRTH_DATE columns to datetime format and standardize to YYYY-MM-DD
users_df['CREATED_DATE'] = pd.to_datetime(users_df['CREATED_DATE']).dt.strftime('%Y-%m-%d')
users_df['BIRTH_DATE'] = pd.to_datetime(users_df['BIRTH_DATE']).dt.strftime('%Y-%m-%d')

In [335]:
# replace null to None
products_df = products_df.where(pd.notnull(products_df), None)
print(products_df.isnull().sum())
users_df = users_df.where(pd.notnull(users_df), None)
print(users_df.isnull().sum())
transactions_df = transactions_df.where(pd.notnull(transactions_df), None)
print(transactions_df.isnull().sum())

CATEGORY_1         111
CATEGORY_2         661
CATEGORY_3       58714
CATEGORY_4      774291
MANUFACTURER    226227
BRAND           226225
BARCODE              0
dtype: int64
ID                      0
CREATED_DATE            0
BIRTH_DATE           3736
STATE                4812
LANGUAGE            30508
GENDER                  0
AGE                  3675
UNREASONABLE_AGE        0
dtype: int64
RECEIPT_ID         0
PURCHASE_DATE      0
SCAN_DATE          0
STORE_NAME         0
USER_ID            0
BARCODE            0
FINAL_QUANTITY     0
FINAL_SALE        55
dtype: int64


# Step 8: Save Cleaned Dataset to CSV
To ensure my cleaned data is ready for visualization in Tableau, I will save the cleaned datasets as CSV files.

In [336]:
# Save the cleaned dataset
products_df.to_csv('/Users/jc/Desktop/xing/fetch/interview/cleaned_products_dataset.csv', index=False, encoding="utf-8")
transactions_df.to_csv('/Users/jc/Desktop/xing/fetch/interview/cleaned_transactions_dataset.csv', index=False, encoding="utf-8")
users_df.to_csv('/Users/jc/Desktop/xing/fetch/interview/cleaned_users_dataset.csv', index=False, encoding="utf-8")

# Summary of Data Quality Issues and Challenges

__Are there any data quality issues present?__

1. Missing Values:
- Users Dataset:
  - BIRTH_DATE: Approximately 3.68% of the values are missing, which could impact age-based analysis.
  - STATE: About 4.81% of the values are missing, potentially affecting regional or geographic segmentation analysis.
  - LANGUAGE: A significant 30.51% of the values are missing, which makes analyzing language preferences challenging.
  - GENDER: Around 5.89% of the values are missing, which could affect demographic segmentation by gender.
- Transactions Dataset:
  - BARCODE: About 11.52% of the values are missing, hindering product-level analysis and integration with the products dataset.
  - FINAL_SALE: Approximately 25% of the values are missing, which could impact revenue-related analysis.
- Products Dataset:
  - CATEGORY_4: A high missing rate of 92% reflects the hierarchical structure of categorization rather than a true data issue.
  - MANUFACTURER and BRAND: Both have about 26.78% missing values, potentially impacting brand or manufacturer-specific analysis.
  - BARCODE: Although the missing rate is only 0.48%, ensuring its accuracy is essential to avoid integration issues.

2. Outliers:
- Users Dataset:
  - BIRTH_DATE: Contains unrealistic values such as January 1, 1900, implying users aged 120+ years. This likely indicates data entry errors.
    - Solution: Implemented a reasonable age range (0–100 years) and flagged unrealistic ages as NaN.
- Transactions Dataset:
  - FINAL_QUANTITY and FINAL_SALE: Contain 0 or blank values, which may indicate invalid transactions. These have been filtered out.

3. Duplicate Records:
- Transactions Dataset:
  - Identified 171 duplicate RECEIPT_ID entries. After cleaning, the dataset has significantly fewer rows, ensuring consistency.
- Products Dataset:
  - Found 57 duplicate records. These rows lack barcodes and may represent the same product under different contexts.
4. Inconsistent Categories:
- Users Dataset:
  - GENDER: Contains 11 unique values, including ambiguous categories such as not_listed and unknown. These values were cleaned and grouped into simplified categories (e.g., female, male, other).
- Products Dataset:
  - The number of unique values varies significantly across category levels. For instance, CATEGORY_1 has 27 categories, while CATEGORY_4 has only 127, reflecting the natural structure of hierarchical categorization.

__Are there any fields that are challenging to understand?__
1. GENDER Field:
- Contains uncommon or redundant values such as non_binary and Non-Binary, requiring further standardization.
- The diversity in values complicates analysis, but grouping them into broader categories improves efficiency.
2. BARCODE Field:
- Missing or duplicate barcodes in the Products and Transactions datasets pose challenges for integration and product-level analysis.
- Certain barcode values (e.g., -1) are illogical and require further cleaning.
3. FINAL_SALE and FINAL_QUANTITY Fields:
- Some transactions are difficult to interpret due to blank or invalid values.
- Proper cleaning requires consultation with stakeholders to ensure alignment with business rules.
- __FINAL_QUANTITY__, in particular, contains decimal values. Upon deeper inspection, the category corresponds to supermarkets, suggesting these could represent items sold by weight (e.g., produce or deli items). However, this assumption requires clarification, and these records have not been removed.
4. CATEGORY_4 Field:
- High missing rate (92%) suggests that many products lack deep-level categorization, limiting detailed product-level analysis.
5. USER_ID and RECEIPT_ID:
- USER_ID: Initially exhibited a many-to-one relationship with transactions (17,694 unique IDs for 50,000 transactions). After cleaning, the dataset has significantly fewer rows, reducing potential biases. However, this highlights the need to validate the cleaned data against expected business logic.
- RECEIPT_ID: Repeated entries may indicate batch transactions or data entry errors. Further investigation is required.
6. Challenge – Limited Transaction Data:
- The Transactions dataset only includes data from June to August 2024, making it insufficient for long-term trend analysis or evaluating seasonal sales patterns.
- Cleaning the data, such as handling duplicates, has further reduced the dataset size. These limitations restrict the ability to derive comprehensive insights into sales performance over time.

# import into MYSQL

import pymysql

#Connect to the MySQL database

connection = pymysql.connect(
        host='localhost',         
        user='root',              
        password='jasmine0418',   
        database='fetch_db'       
    )
cursor = connection.cursor()

#Load the cleaned dataset 

csv_file_path = '/Users/jc/Desktop/xing/fetch/interview/cleaned_products_dataset.csv'
df = pd.read_csv(csv_file_path, dtype=str)

#replace null to None

df = df.where(pd.notnull(df), None)
print(df.isnull().sum())

for _, row in df.iterrows():
    data = (
        row['CATEGORY_1'],
        row['CATEGORY_2'],
        row['CATEGORY_3'],
        row['CATEGORY_4'],
        row['MANUFACTURER'],
        row['BRAND'],
        row['BARCODE']
    )
    print(data)  # Debug: Print data to verify each row's content
    cursor.execute("""
        INSERT INTO products (CATEGORY_1, CATEGORY_2, CATEGORY_3, CATEGORY_4, MANUFACTURER, BRAND, BARCODE) 
        VALUES (%s, %s, %s, %s, %s, %s, %s)
    """, data)
connection.commit()

csv_file_path = '/Users/jc/Desktop/xing/fetch/interview/cleaned_transactions_dataset.csv'
df2 = pd.read_csv(csv_file_path, dtype=str)

#Replace null values and remove rows with missing BARCODE

df2 = df2.where(pd.notnull(df2), None)

#Truncate the `transactions` table

print("Truncating the transactions table...")
cursor.execute("TRUNCATE TABLE transactions;")
connection.commit()
print("Transactions table truncated successfully.")

#Insert the cleaned data

print("Inserting cleaned data into transactions table...")
sql = """
    INSERT INTO transactions (RECEIPT_ID, PURCHASE_DATE, SCAN_DATE, STORE_NAME, USER_ID, BARCODE, FINAL_QUANTITY, FINAL_SALE)
    VALUES (%s, %s, %s, %s, %s, %s, %s, %s)
"""
for _, row in df2.iterrows():
    data = (
        row['RECEIPT_ID'],
        row['PURCHASE_DATE'],
        row['SCAN_DATE'],
        row['STORE_NAME'],
        row['USER_ID'],
        row['BARCODE'],
        row['FINAL_QUANTITY'],
        row['FINAL_SALE']
    )
    cursor.execute(sql, data)

#Commit changes and close the connection

connection.commit()

#Load the cleaned dataset

csv_file_path = '/Users/jc/Desktop/xing/fetch/interview/cleaned_users_dataset.csv'
df3 = pd.read_csv(csv_file_path, dtype=str)

#Replace null values with None

df3 = df3.where(pd.notnull(df3), None)
print(f"Total rows in cleaned DataFrame: {len(df3)}")

#Truncate the `users` table

print("Truncating the users table...")
cursor.execute("TRUNCATE TABLE users;")
connection.commit()
print("Users table truncated successfully.")

#Insert the cleaned data into `users` table

print("Inserting cleaned data into users table...")
sql = """
    INSERT INTO users (ID, CREATED_DATE, BIRTH_DATE, STATE, LANGUAGE, GENDER, AGE, UNREASONABLE_AGE)
    VALUES (%s, %s, %s, %s, %s, %s, %s, %s)
"""

#Loop through the DataFrame and insert data

for _, row in df3.iterrows():
    data = (
        row['ID'],
        row['CREATED_DATE'],
        row['BIRTH_DATE'],
        row['STATE'],
        row['LANGUAGE'],
        row['GENDER'],
        row['AGE'],
        row['UNREASONABLE_AGE']
    )
    cursor.execute(sql, data)

#Commit the changes

connection.commit()
print("Cleaned data inserted into users table successfully.")

cursor.close()
connection.close()